libbeat/publisher/queue/memqueue/ringbuf.go

```diff
@@ -0,0 +1,293 @@
+package memqueue
+
+import (
+	"fmt"
+
+	"github.com/elastic/beats/libbeat/publisher"
+)
+
+// Internal event ring buffer.
+// The ring is split into 2 regions.
+// Region A contains active events to be send to consumers, while region B can
+// only be filled by producers, if there is no more space in region A. Splitting
+// the ring buffer into regions enables the broker to send batches of type
+// []publisher.Event to the consumer without having to copy and/or grow/shrink the
+// buffers.
+type ringBuffer struct {
+	buf eventBuffer
+
+	regA, regB region
```

Comment:
Just for the record: I think `regA` and `regB` could be renamed to `regConsumer` and `regProducer`, so the could can make more sense without the comments. But as it is just moved from a different a file I can live with it. It is already an improvement over the previous state.

Reply:
tl;dr: Only purpose of regB is to handle the filling the ring buffers when the indices in the ring buffer 'wrap around'.

regA is regConsumer and regProducer as long as there is space left in regA :)
regB only comes into existence if the ring buffer wraps around (no more space in regA).
Once regB exists, regA is indeed consumer only and regB is producer only.
After wrapping around regB becomes regA, once regA becomes empty. From then on regA is again shared by producers and consumers until another 'wrap around' on the ring buffer.

Reply:
Yes, I have seen that. I was just trying to come up with names which could help the understanding more. But ofc, the proposed names were not descriptive either. :(